op-chain-ops/cmd/check-canyon/main.go

```diff
@@ -0,0 +1,278 @@
+		return err
+	}
+
+	if canyonActive && block.Withdrawals() == nil {
+		return errors.New("No withdrawals in a canyon block")
+	} else if !canyonActive && block.Withdrawals() != nil {
+		return errors.New("Withdrawals in a pre-canyon block")
+	}
+	return nil
+}
+
+func ValidateCreate2Deployer(ctx Args, canyonActive bool) error {
+	addr := common.HexToAddress("0x13b0D85CcB8bf860b6b79AF3029fCA081AE9beF2")
+	code, err := ctx.Client.CodeAt(context.Background(), addr, new(big.Int).SetUint64(ctx.Number))
+	if err != nil {
+		return err
+	}
+	codeHash := crypto.Keccak256Hash(code)
+	expectedCodeHash := common.HexToHash("0xb0550b5b431e30d38000efb7107aaa0ade03d48a7198a140edda9d27134468b2")
+
+	if canyonActive && codeHash != expectedCodeHash {
+		return fmt.Errorf("Canyon active but code hash does not match. have: %v, want: %v", codeHash, expectedCodeHash)
+	} else if !canyonActive && codeHash == expectedCodeHash {
+		return fmt.Errorf("Canyon not active but code hashes do match. codeHash: %v", codeHash)
+	}
+
+	return nil
+}
+
+// CheckActivation takes a function f which determines in a specific block follows the rules of a fork.
+// forkActivated tells `f` if the fork is active or not. `f` is called twice: First to validate that
+// there is no error when checking the new value and second to validate the it returns an error when
+// attempting to validate the block against the opposite of what is is.
+// If any error is encounted, valid is set to false.
+func CheckActivation(f func(Args, bool) error, ctx Args, forkActivated bool, validatePreState bool, valid *bool) {
+	if forkActivated {
+		if err := f(ctx, true); validatePreState && err != nil {
+			log.Error("Pre-state was invalid when it was expected to be valid", "err", err)
+			*valid = false
+		}
+		if err := f(ctx, false); err == nil {
+			log.Error("Post-state was valid when it was expected to be invalid")
+			*valid = false
+		}
+	} else {
+		if err := f(ctx, true); validatePreState && err == nil {
+			log.Error("Pre-state was valid when it was expected to be invalid")
+			*valid = false
+		}
+		if err := f(ctx, false); err != nil {
+			log.Error("Post-state was invalid when it was expected to be valid", "err", err)
+			*valid = false
+		}
+	}
+}
+
+func main() {
+	logger := log.New()
+
+	// Define the flag variables
+	var (
+		canyonActive bool
+		number       uint64
+		elasticity   uint64
+		rpcURL       string
+	)
+
+	valid := true
+
+	// Define and parse the command-line flags
+	flag.BoolVar(&canyonActive, "canyon", false, "Set this flag to assert canyon behavior")
+	flag.Uint64Var(&number, "number", 31, "Block number to check")
+	flag.Uint64Var(&elasticity, "elasticity", 6, "Specify the EIP-1559 elasticity. 6 on mainnet/sepolia. 10 on goerli")
+	flag.StringVar(&rpcURL, "rpc-url", "http://localhost:8545", "Specify the L2 ETH RPC URL")
+
+	// Parse the command-line arguments
+	flag.Parse()
+
+	l1RPC, err := client.NewRPC(context.Background(), logger, rpcURL, client.WithDialBackoff(10))
```

Comment:
Is the rpcURL for L1 or L2?

Reply:
L2 url (but treating it more similar to the L1). Will fix